### ISSUE TITLE
Pre-create check: "Hyper-V PowerShell Module is not available"

### ISSUE THREAD

**Issue by pldmgg** [docker/machine#4424]
Please note: this is NOT a duplicate of #4342.

Running docker-for-windows 18.03.0.16511-edge.

Command:

    PS C:\Users\pdadmin> docker-machine --debug create -d hyperv --hyperv-virtual-switch "ToInternal" DockerHostA
    Docker Machine Version:  0.14.0, build 89b8332
    Found binary path at C:\Program Files\Docker\Docker\Resources\bin\docker-machine.exe
    Launching plugin server for driver hyperv
    Plugin server listening at address 127.0.0.1:49951
    () Calling .GetVersion
    Using API Version  1
    () Calling .SetConfigRaw
    () Calling .GetMachineName
    (flag-lookup) Calling .GetMachineName
    (flag-lookup) Calling .DriverName
    (flag-lookup) Calling .GetCreateFlags
    Found binary path at C:\Program Files\Docker\Docker\Resources\bin\docker-machine.exe
    Launching plugin server for driver hyperv
    Plugin server listening at address 127.0.0.1:49955
    () Calling .GetVersion
    Using API Version  1
    () Calling .SetConfigRaw
    () Calling .GetMachineName
    (DockerHostA) Calling .GetMachineName
    (DockerHostA) Calling .DriverName
    (DockerHostA) Calling .GetCreateFlags
    (DockerHostA) Calling .SetConfigFromFlags
    Reading certificate data from C:\Users\pdadmin\.docker\machine\certs\ca.pem
    Decoding PEM data...
    Parsing certificate...
    Reading certificate data from C:\Users\pdadmin\.docker\machine\certs\cert.pem
    Decoding PEM data...
    Parsing certificate...
    Running pre-create checks...
    (DockerHostA) Calling .PreCreateCheck
    (DockerHostA) DBG | [executing ==>] : C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoProfile -NonInteractive @(Get-Command hyper-v\Get-VM).ModuleName
    (DockerHostA) DBG | [stdout =====>] : hyper-v
    (DockerHostA) DBG |
    (DockerHostA) DBG | [stderr =====>] :
    Error with pre-create check: "Hyper-V PowerShell Module is not available"
    notifying bugsnag: [Error with pre-create check: "Hyper-V PowerShell Module is not available"]

The issue is again with https://github.com/docker/machine/blob/master/drivers/hyperv/powershell.go with the `hypervAvailable()` function.

As you can see by the debug above, trying this in PowerShell...

    & "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -NonInteractive "@(Get-Command hyper-v\Get-VM).ModuleName"

...gives you lowercase, i.e. 'hyper-v' and the Go code checks for uppercase 'Hyper-V'.

There are a few solutions.

BEST SOLUTION (in my opinion):

Generally, when determining if a particular PowerShell Module is installed/available, you would use the `Get-Module` cmdlet. So line 58 in https://github.com/docker/machine/blob/master/drivers/hyperv/powershell.go would go from...
    
    stdout, err := cmdOut("@(Get-Command hyper-v\\Get-VM).ModuleName")

...to...

    stdout, err := cmdOut("@(Get-Module -ListAvailable hyper-v).Name | Get-Unique")

SOLUTION 2:

Change line 58 from...
    
    stdout, err := cmdOut("@(Get-Command hyper-v\\Get-VM).ModuleName")

...to...

    stdout, err := cmdOut("@(Get-Command Hyper-V\\Get-VM).ModuleName")

SOLUTION 3:

Make line 64...

    if resp[0] != "Hyper-V" {

...case insensitive.

**Comment by ramit-rulz** [comment]
I am getting the same issue on Windows Server 2016 VM. Is there a timeline for this fix to be released? Is there a workaround until then?

**Comment by scmikes** [comment]
Any workaround for this on windows 10?

**Comment by Chrisly3ear** [comment]
Same Problem. I think the solution should be

`stdout, err := cmdOut("@(Get-Module -ListAvailable hyper-v).Name | Get-Unique")`

like OP suggested. The same issue exists in Minikube, btw.

**Comment by pldmgg** [comment]
@scmikes 
Workaround is to just install your own Debian/Ubuntu/CentOS/RedHat VM, create an account called 'docker' with sudo privileges, setup ssh public key auth, and do:

    docker-machine create --driver generic --generic-ip-address $VMIPv4Address --generic-ssh-user docker --generic-ssh-key "$("$HOME\.ssh\private_key.pem" -replace "\\","/")" $VMName

I strongly recommend doing this regardless of whether or not this particular bug gets fixed because of this issue that I ran into:

https://github.com/boot2docker/boot2docker/issues/1311

**Comment by ckadluba** [comment]
Line 58 in https://github.com/docker/machine/blob/master/drivers/hyperv/powershell.go (see above).

**Comment by ajtortolero** [comment]
I understood that, I can not locate the file on my machine.

**Comment by ajtortolero** [comment]
I understood that, I can not locate the file on my machine.

**Comment by samirsyed** [comment]
@ramit-rulz A workaround is to downgrade docker-machine to [v0.13.0](https://github.com/docker/machine/releases/tag/v0.13.0)

1. Download [32-bit](https://github.com/docker/machine/releases/download/v0.13.0/docker-machine-Windows-i386.exe) or [64-bit](https://github.com/docker/machine/releases/download/v0.13.0/docker-machine-Windows-x86_64.exe) version 
2. Replace _C:\Program Files\Docker\Docker\resources\bin\docker-machine.exe_ with the newly downloaded file.

**Comment by scmikes** [comment]
Just wanted to verify that I was able to create new docker machine on windows 10 using downgrade to v0.13.0.  Thank you very much samirsyed for posting instructions

**Comment by n1ge1m** [comment]
samirsyed - that worked for me

**Comment by borkke** [comment]
Same problem here...
>docker-machine -version => version 0.14.0, build 89b8332

After the downgrade, it works fine

**Comment by goudarzi** [comment]
+1 (@samirsyed's downgrade solution worked for me too. 14 > 13)

**Comment by jhenrypp** [comment]
It did install, but it gave me Error creating machine: Error in driver during machine creation: exit status 1 at last.

**Comment by Ouro17** [comment]
Will docker-machine version 13 update in the next update of docker or will be need to do it manually like the downgrade?
Thank you all.

**Comment by scmikes** [comment]
Is the status of this bug correct?  It looks like a fix was submitted.  At least one of the comments has a merge tag

**Comment by surajsubs** [comment]
I am having the same issue as reported by @andrewsav-datacom.
![image](https://user-images.githubusercontent.com/8228017/38656256-30faf098-3dce-11e8-9d94-562b6e836e12.png)

It appears that when module name is retrieved in powershell run in non-administrative mode, the module name is printed in lower case as seen below

![image](https://user-images.githubusercontent.com/8228017/38656335-ab12bc08-3dce-11e8-8cdf-7e61832d400d.png)

I have downgraded minicube to 0.25.2 and I no longer see the issue.

**Comment by gbraad** [comment]
> SOLUTION 3:

I would suggest to perform a `strings.ToLower()`. As this guarantees going forward in an 'case insensitive'- way. However, the fix already exists, but is not tagged.

@beppler @shin- 
The version tagged as `0.14.0` does works, is NOT tagged: See, https://github.com/docker/machine/pull/4426 If you pull checksum: e501d1bcb54f09de8ff71aa20010321d18425d19 using `Godep` it would work.

**Comment by beppler** [comment]
@andrewsav-datacom, most likely your profile configuration is explicit importing the Hyper-V module using the following command.

```powershell
Import-Module Hyper-V
```

The case of module name is controlled by the way you explicit import (or not import) the Hyper-V module.

If you don't explicit import it (like when you execute from cmd as you use the `-NoProfile` switch) it will report hyper-v.

If you explict import the Hyper-V module, powershell will report it with the same case you use on Import-Module.

For example. My profile does not explicit import Hyper-V module so if i run:

```powershell
Import-Module hyper-V
(Get-Command hyper-v\get-vm).Module.Name
```

Module name is reported as: `hyper-V`.

**Comment by gbraad** [comment]
right... but this behaviour also happens when the module is actually loaded. The `Import-Module` seems to override the value.

Anyways, I believe solution 3 is cleaner for future use... but the current fix will do

**Comment by gbraad** [comment]
I also confirmed this earlier today ...
![image](https://user-images.githubusercontent.com/1894/38676205-ddcd7d54-3e8c-11e8-8002-11b1618bcf17.png)

**Comment by MikyChow** [comment]
@gbraad  try docker --debug create -d hyperv yourmachinename,then get the debug info.

**Comment by Jishun** [comment]
I'm able to manipulate the module name shown with `get-command hyper-v\get-vm`  by `remove-module` then 'import-module' in the case that I want 
but whichever case I tried, it docker-machine gives this same error

**Comment by MikyChow** [comment]
@samirsyed  Thanks,That worked for me!

**Comment by gbraad** [comment]
@MikyChow no need to collect this. We already know what the issue is: https://github.com/docker/machine/blob/1e3b6b64504f2970252b06119d6dc1fb1982b4b4/drivers/hyperv/powershell.go#L58 and  it's OK for us (minikube, minishift) now. The check works... but the release as 0.14.1 or whatever was never made, so pulling docker/machine as dependency has to be done on a sha hash and this is far from ideal.

**Comment by kierenj** [comment]
How can I downgrade docker-machine on Windows, please?  Alternatively, is there a pending hotfix release?

**Comment by dlakatos847** [comment]
+1

**Comment by cedric05** [comment]
+1

**Comment by gbraad** [comment]
Off-topic discussion; please continue this by email.

**Comment by LeoTSH** [comment]
Hi There, may I ask if this is still an ongoing issue? I'm on 18.03.1-ce and encountering the issue. I have tried Import-Module Hyper-V but the issue persists.

**Comment by AndrewSav** [comment]
Well it's marked as open, you encountered it lately, so we pretty much can conclude that it's not fixed yet.

**Comment by gbraad** [comment]
Still, as no new tagged release has been made.

Both minikube and minishift pulled a later version (after v1.4.0) and do
not have this issue

**Comment by SpamhaterAlex** [comment]
I dont understand, what you are writing, but I could not see a solution? 
As I see above, I have to change something in code of powershell? But how? 

I get the same problem above with 

Error with pre-create check: "Hyper-V PowerShell Module is not available"

**Comment by gbraad** [comment]
TL;dr: **Solution is available inthe source, but has not been made into a deliverable.**

The guys responsible for this repository have to make a release, simple as that. And oas of now, they haven't done this (although several requests have been made).

**Comment by SpamhaterAlex** [comment]
Ok @gbraad  can you pls tell me how I can use this "source"?

**Comment by SpamhaterAlex** [comment]
Or integrate it, to solve the problem above

**Comment by gbraad** [comment]
workaround: `Remove-Module hyper-v; Import-Module Hyper-V` but haven't checked this myself... might work

or
```
git checkout http://github.com/docker/machine.git
cd machine
``` 
note: which is the latest...

or do at least a `git checkout e501d1b`

**Comment by SpamhaterAlex** [comment]
Thanks, so only what I need to do, is 

`git checkout e501d1b `

and then I am able to write the command above? 👍

**Comment by gbraad** [comment]
Eh... wish it was that simple

To be honest, you should request a `v.0.14.x` release from the maintainers, such as @shin-

**Comment by SpamhaterAlex** [comment]
When I run my git cmd with this command, 

I get this: 

`fatal: Not a git repository (or any of the parent directories): .git`

**Comment by SpamhaterAlex** [comment]
@gbraad okay xD

**Comment by stormxing** [comment]
+1

**Comment by juliolima86** [comment]
+1

**Comment by cedric05** [comment]
+1

**Comment by nickreynke** [comment]
+1

**Comment by MisterPuffyPants** [comment]
+1

**Comment by jplonghi** [comment]
+1

**Comment by anikanovna** [comment]
+1

**Comment by crowdy** [comment]
+1

**Comment by mdschweda** [comment]
+1

**Comment by rand0mize71** [comment]
+1

**Comment by gopikishanm** [comment]
+1

**Comment by potierv** [comment]
+1

**Comment by kantenkugel** [comment]
Guys, can you please stop +1ing? If you want to show your support, just add an reaction or w/e.

This doesn't boost the speed of implementation at all. Just annoys people subscribed to the thread to see when actual updates happen.

**Comment by AndrewSav** [comment]
@kantenkugel, nah, that's fine. Does not annoy _me_.

**Comment by dasmedium** [comment]
Any word on getting this release out?

**Comment by gbraad** [comment]
The time it takes is telling for the maintenance and support of this library. We have already forked this as part of docker-machines/machine, and are considering to do our own releases instead.

**Comment by CharlesStover** [comment]
Verifying for anyone also finding this via Google that downgrading from 0.14 to 0.13 resolves this issue. Practically pulled my hair out trying to get the docker machine to start. Thanks @samirsyed.

**Comment by gmsantos** [comment]
Downgrading to 0.13 didn't worked for me, but building from master branch worked well.

See [CONTRIBUTING.md](https://github.com/docker/machine/blob/master/CONTRIBUTING.md ) for more details on how to build from source. 

I hope docker maintainers release a new version soon.

**Comment by saifjsl** [comment]
+1

**Comment by NikitaZhevnitskiy** [comment]
+1

**Comment by abigailarreguin** [comment]
I had this same issue for awhile. I installed the docker toolbox and that's all it took to work. I already met all the requirements so I am guessing there were specific versions I needed that were within the toolbox. I am all set up and running now. Hope this helps!

**Comment by m-suizu** [comment]
```
PS C:\> docker-machine --version
docker-machine.exe version 0.14.0, build 89b8332

PS C:\> Import-Module Hyper-v
PS C:\> (Get-Command hyper-v\get-vm).Module.Name
Hyper-v

PS C:\> docker-machine --debug create -d hyperv --hyperv-virtual-switch dockerswitch vm

...

Running pre-create checks...
(local-squid) Calling .PreCreateCheck
(local-squid) DBG | [executing ==>] : C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoProfile -NonInteractive @(Get-Command hyper-v\Get-VM).ModuleName
(local-squid) DBG | [stdout =====>] : hyper-v
(local-squid) DBG |
(local-squid) DBG | [stderr =====>] :
Error with pre-create check: "Hyper-V PowerShell Module is not available"
```

**Comment by nigel-dewar** [comment]
Doesnt work. switching back to kubernetes.

**Comment by q587p** [comment]
Microsoft Windows [Version 10.0.17134.48]
Linux q587p-work 4.4.0-17134-Microsoft №48-Microsoft Fri Apr 27 18:06:00 PST 2018 x86_64 x86_64 x86_64 GNU/Linux
Docker version 18.03.1-ce, build 9ee9f40

Downgrade to 0.13 helps with

`Error with pre-create check: "Hyper-V PowerShell Module is not available"`

But now I have next problem:

`Error with pre-create check: "Hyper-v commands have to be run as an Administrator"`

If I used .exe then this is simple way to change this (C:\Program Files\Docker\Docker\resources\bin\docker-machine.exe → right click and select Compatibility and check the box "Run this program as an administrator").

But how and where change this permission in WSL?..

**Comment by kantenkugel** [comment]
@q587p Just run your bash / shell / command line as administrator. That should fix it. WSL might already work with `sudo`

**Comment by q587p** [comment]
Thnx @kantenkugel! It wasn't obvious.

Next one appears:

`Error creating machine: Error in driver during machine creation: open /root/.docker/machine/machines/myvm1/fixed.vhd: no such file or directory`

**Comment by Yan-Matskevich** [comment]
Have the same issue with "Hyper-V PowerShell Module is not available" error and downgrade to version 0.13 doesn't solve the problem ( with old one a have another problem). As a workaround I use Oracle VM virtualbox now. So is there any news about new docker-machine version with resolved issue :)

**Comment by shin-** [comment]
https://github.com/docker/machine/releases/tag/v0.15.0

Thank you for your patience.

**Comment by ianisms** [comment]
I have the same issue with 0.15: Error with pre-create check: "Hyper-V PowerShell Module is not available"
winver: 1803 (OS Build 17692.1000)
docker-machine version 0.15.0, build b48dc28d

**Comment by webstandardcss** [comment]
#Here is the solution

## https://github.com/docker/machine/releases/download/v0.15.0/docker-machine-Windows-x86_64.exe

### Save the file to your existing directory containing `docker-machine.exe`.   Rename the filename to `docker-machine.exe` and replace the old file with this one.

For my system this is the location for docker-machine.exe
`/c/Program Files/Docker/Docker/Resources/bin/docker-machine.exe`

### Build Instructions

1. Create virtual switch in Hyper-V manager named myswitch
2. Request docker to create a VM named myvm1

`docker-machine create -d hyperv --hyperv-virtual-switch "myswitch" myvm1`

**Comment by mludi** [comment]
Awesome. Thank you @webstandardcss

**Comment by programmingwithalex** [comment]
To get the `docker-machine` command working on PowerShell I followed a similar approach to @webstandardcss.

1. Went to the `docker-machine` release page: https://github.com/docker/machine/releases/tag/v0.16.2
2. Scroll to the bottom and download the latest executable for Windows (currently named `docker-machine-Windows-x86_64.exe`)
3. *Rename this file* to `docker-machine.exe`
4. Move this file to `C:\Program Files\Docker\Docker\resources\bin`

I was then able to run `docker-machine` in my PowerShell prompt.
